libsolidity/codegen/CompilerContext.cpp

```diff
@@ -556,32 +538,13 @@ LinkerObject const& CompilerContext::assembledObject() const
 	return object;
 }
 
-FunctionDefinition const& CompilerContext::resolveVirtualFunction(
-	FunctionDefinition const& _function,
-	vector<ContractDefinition const*>::const_iterator _searchStart
-)
+ContractDefinition const* CompilerContext::superContract(ContractDefinition const& _contract) const
 {
-	string name = _function.name();
-	FunctionType functionType(_function);
-	auto it = _searchStart;
-	for (; it != m_inheritanceHierarchy.end(); ++it)
-		for (FunctionDefinition const* function: (*it)->definedFunctions())
-			if (
-				function->name() == name &&
-				!function->isConstructor() &&
-				FunctionType(*function).asCallableFunction(false)->hasEqualParameterTypes(functionType)
-			)
-				return *function;
-	solAssert(false, "Super function " + name + " not found.");
-	return _function; // not reached
-}
-
-vector<ContractDefinition const*>::const_iterator CompilerContext::superContract(ContractDefinition const& _contract) const
-{
-	solAssert(!m_inheritanceHierarchy.empty(), "No inheritance hierarchy set.");
-	auto it = find(m_inheritanceHierarchy.begin(), m_inheritanceHierarchy.end(), &_contract);
-	solAssert(it != m_inheritanceHierarchy.end(), "Base not found in inheritance hierarchy.");
-	return ++it;
+	solAssert(m_mostDerivedContract, "Most derived contract not set.");
+	auto const& hierarchy = m_mostDerivedContract->annotation().linearizedBaseContracts;
+	auto it = find(hierarchy.begin(), hierarchy.end(), &_contract);
+	solAssert(it != hierarchy.end(), "Base not found in inheritance hierarchy.");
+	return *(++it);
```

Comment:
I only did a superficial look here. But is `return *(++it)` really correct? what is happening, if `it` is already pointing to the last element?

Reply:
Yep, that was exactly the bug :)